odoo/orm/fields_relational.py

```diff
@@ -1054,7 +1054,7 @@ def write_real(self, records_commands_list, create=False):
                 self.read(records.browse(missing_ids))
 
         # determine new relation {x: ys}
-        old_relation = {record.id: set(record[self.name]._ids) for record in records}
+        old_relation = {record.id: set() if create else set(record[self.name]._ids) for record in records}
```

Comment:
Nice catch, can you add a test for it (query count) ? It saves us one query right?

Also, I don't really understand how it is related with the cache pollution :thinking: ?

Reply:
It is during the creation and everything is already prefetched, I think it is the same amount of queries. From what I can remember, the failing test was trying to access `record[self.name]` but did not have the permissions to do so (a kind of test where you can create but cannot read).
Anyways, during creation there are no old values (by definition).

Reply:
I get it. The cache is always set to an empty tuple in case of creation before reaching this code, it is why you don't save a query (see https://github.com/odoo/odoo/blob/3445c96828a954feceb0ba2d9874931b3a87f96e/odoo/orm/models.py#L5068)

> From what I can remember, the failing test was trying to access record[self.name] but did not have the permissions to do so (a kind of test where you can create but cannot read).

Then you don't really fix the issue, `write` still won't work for no reason. If you fix the cache poisoning, this fix should change by reading in sudo the old relation `record.sudo()[self.name]._ids` instead. Then IMO, this change doesn't make sense.

Reply:
Reading with sudo bypasses permissions.
When you write, you have read the previous values, so you have access. When creating, you may not have additional permissions and the old values do not make sense when creating a record.

Reply:
> Reading with sudo bypasses permissions.

It is my point, if you fix the cache poisoning, you can bypasses read permission to write the X2many. The permission should be check when you get the X2many from the cache anyway.

> When you write, you have read the previous values, so you have access.

Write doesn't implies that you have read access, right? In the same way that you can create without the access to read. Or I miss something ?